src/halmos/__main__.py

```diff
@@ -397,6 +408,279 @@ def is_global_fail_set(context: CallContext) -> bool:
+
+                    print("\nTrace:")
+                    render_trace(post_ex.context)
+
+                # stop if no more invariants to test
+                if not funsigs:
+                    return next_exs, funsigs
+
+                # update the frontier states to be returned.
+                # NOTE: this state excludes any changes made during the execution of invariant test functions to prevent unnecessary increases in path condition complexity.
+                next_exs.append(post_ex)
+
+    return next_exs, funsigs
+
+
+def run_target_contract(ctx: ContractContext, ex: Exec, addr: Address) -> list[Exec]:
+    """
+    Executes a given contract from a given input state and returns all output states.
+
+    Args:
+        ctx: The context of the test contract, which differs from the target contract to be executed.
+        ex: The input state.
+        addr: The address of the contract to be executed.
+
+    Returns:
+        A list of output states.
+
+    Raises:
+        ValueError: If the contract name cannot be found for the given address.
+    """
+    args = ctx.args
+
+    # retrieve the contract name and metadata from the given address
+    code = ex.code[addr]
+    contract_name = code.contract_name
+    filename = code.filename
+
+    if not contract_name:
+        raise ValueError(f"couldn't find the contract name for: {addr}")
+
+    contract_json = BuildOut().get_by_name(contract_name, filename)
+    abi = get_abi(contract_json)
+    method_identifiers = contract_json["methodIdentifiers"]
+
+    results = []
+
+    # iterate over each function in the target contract
+    for fun_sig, fun_selector in method_identifiers.items():
+        fun_name = fun_sig.split("(")[0]
+        fun_info = FunctionInfo(fun_name, fun_sig, fun_selector)
+
+        # skip if 'pure' or 'view' function that doesn't change the state
+        state_mutability = abi[fun_sig]["stateMutability"]
+        if state_mutability in ["pure", "view"]:
+            if args.debug:
+                print(f"Skipping {fun_name} ({state_mutability})")
+            continue
+
+        try:
+            # initialize symbolic execution environment
+            sevm = SEVM(args, fun_info)
+            # TODO: reuse solver across different functions
+            solver = mk_solver(args)
+            path = Path(solver)
+            path.extend_path(ex.path)
+
+            # prepare calldata and dynamic parameters
+            cd, dyn_params = mk_calldata(
+                abi, fun_info, args, new_symbol_id=ex.new_symbol_id
+            )
+            path.process_dyn_params(dyn_params)
+
+            # create a symbolic message value
+            msg_value = BitVec(
+                f"msg_value_{id_str(addr)}_{uid()}_{ex.new_symbol_id():>02}",
+                BitVecSort256,
+            )
+
+            # construct the transaction message
+            message = Message(
+                target=addr,
+                caller=ex.this(),
```

Comment:
I think we probably want a symbolic sender here -- maybe just avoiding address(0) and known contract addresses?

foundry behavior:

> Target Senders: The invariant test fuzzer picks values for msg.sender at random when performing fuzz campaigns to simulate multiple actors in a system by default. If desired, the set of senders can be customized in the setUp function.

Reply:
symbolic sender and origin: https://github.com/a16z/halmos/pull/454/commits/de984d1897223db613b3a19f9ac1edef7c96bff9

Reply:
support for finer control of msg.sender was added in the todo list